services/gossip/adapter/direct_transport_test.go

```diff
@@ -0,0 +1,278 @@
+		require.NoError(t, err, "test peer could not write to local transport")
+		require.Equal(t, len(buffer), written)
+
+		buffer = []byte{0} // dummy buffer just to see when the connection closes
+		_, err = h.peerTalkerConnection.Read(buffer)
+		require.Error(t, err, "test peer should be disconnected from local transport")
+
+		h.verifyTransportListenerNotCalled(t)
+	})
+}
+
+func TestDirectIncoming_TransportListenerDoesNotReceiveCorruptData_PayloadSize(t *testing.T) {
+	t.Parallel()
+	test.WithContext(func(ctx context.Context) {
+
+		h := newDirectHarnessWithConnectedPeers(t, ctx)
+		defer h.cleanupConnectedPeers()
+
+		h.transport.RegisterListener(h.listenerMock, nil)
+		h.expectTransportListenerNotCalled()
+
+		buffer := []byte{0x01, 0x00, 0x00, 0x00, 0x99, 0x99, 0x99, 0x99, 0x11, 0x00, 0x00, 0x00}
+		written, err := h.peerTalkerConnection.Write(buffer)
+		require.NoError(t, err, "test peer could not write to local transport")
+		require.Equal(t, len(buffer), written)
+
+		buffer = []byte{0} // dummy buffer just to see when the connection closes
+		_, err = h.peerTalkerConnection.Read(buffer)
+		require.Error(t, err, "test peer should be disconnected from local transport")
+
+		h.verifyTransportListenerNotCalled(t)
+	})
+}
+
+func TestDirectOutgoing_SendsKeepAliveWhenNothingToSend(t *testing.T) {
+	test.WithContext(func(ctx context.Context) {
+
+		h := newDirectHarnessWithConnectedPeers(t, ctx)
+		defer h.cleanupConnectedPeers()
+
+		for numKeepAliveSent := 0; numKeepAliveSent < 2; numKeepAliveSent++ {
+			data, err := h.peerListenerReadTotal(1, 4)
+			require.NoError(t, err, "test peer server could not read keepalive from local transport")
+			require.Equal(t, []byte{0x00, 0x00, 0x00, 0x00}, data) // keepalive content (zero payloads)
+		}
+	})
+}
+
+func TestDirectIncoming_TransportListenerIgnoresKeepAlives(t *testing.T) {
+	test.WithContext(func(ctx context.Context) {
+
+		h := newDirectHarnessWithConnectedPeers(t, ctx)
+		defer h.cleanupConnectedPeers()
+
+		h.transport.RegisterListener(h.listenerMock, nil)
+		h.expectTransportListenerCalled([][]byte{{0x11}, {0x22, 0x33}})
+
+		for numKeepAliveReceived := 0; numKeepAliveReceived < 2; numKeepAliveReceived++ {
+			buffer := []byte{0x00, 0x00, 0x00, 0x00} // keepalive content (zero payloads)
+			written, err := h.peerTalkerConnection.Write(buffer)
+			require.NoError(t, err, "test peer could not write to local transport")
+			require.Equal(t, len(buffer), written)
+		}
+
+		buffer := []byte{0x02, 0x00, 0x00, 0x00, 0x01, 0x00, 0x00, 0x00, 0x11, 0x00, 0x00, 0x00, 0x02, 0x00, 0x00, 0x00, 0x22, 0x33, 0x00, 0x00}
+		written, err := h.peerTalkerConnection.Write(buffer)
+		require.NoError(t, err, "test peer could not write to local transport")
+		require.Equal(t, len(buffer), written)
+
+		h.verifyTransportListenerCalled(t)
+	})
+}
+
+// wanted to simulate Timeout on Send instead of Error but was unable to
+func TestDirectOutgoing_ErrorDuringSendCausesReconnect(t *testing.T) {
```

Comment:
I did not understand what you're testing here

Reply:
that when the adapter is sending data to a peer and the send fails due to a socket error (or a socket timeout) - the adapter disconnects and reconnects the outgoing socket